dsniff.8
.TH DSNIFF 8 "10 December 2025" "2.5" "Dsniff man page"
.ad
.fi
.SH NAME
dsniff
\-
password sniffer
.SH SYNOPSIS
.na
.nf
.fi
\fBdsniff\fR [\fB-v\fR] [\fB-c\fR] [\fB-a\fR] [\fB-d\fR] [\fB-m\fR] [\fB-D\fR] [\fB-N\fR] [\fB-P\fR] [\fB-i
\fIinterface\fR | \fB-p \fIpcapfile\fR] [\fB-s \fIsnaplen\fR] [\fB-M \fImagics\fR] [\fB-f \fIservices\fR]
[\fB-t \fItrigger[,...]\fR]
[\fIexpression\fR]
.SH DESCRIPTION
.ad
.fi
\fBdsniff\fR is a password sniffer that handles FTP, Telnet, SMTP, HTTP, HTTPS,
POP, poppass, NNTP, IMAP, SNMP, LDAP, Rlogin, RIP, OSPF, PPTP,
MS-CHAP, NFS, VRRP, YP/NIS, SOCKS, X11, CVS, IRC,
PostgreSQL, Meeting Maker, Citrix ICA, Microsoft SMB, Oracle SQL*Net, Sybase, and Microsoft SQL
protocols.
.LP
\fBdsniff\fR automatically detects and minimally parses each
application protocol, saving only the interesting bits, logging only unique
authentication attempts. Full TCP/IP reassembly is provided by
libnids(3).
.LP
I wrote \fBdsniff\fR with honest intentions—to audit my own network
and to demonstrate the insecurity of cleartext network protocols.
Please do not abuse this software.
.SH OPTIONS
.IP \fB-c\fR
Perform half-duplex TCP stream reassembly to handle asymmetrically
routed traffic (such as when using arpspoof(8) to intercept client
traffic bound for the local gateway).
.IP \fB-v\fR
Verbose mode. Show banners.
.IP \fB-a\fR
Show all results. Otherwise, dsniff attempts to suppress duplicate entries.
.IP \fB-d\fR
Enable debugging mode.
.IP \fB-m\fR
Force protocol detection/DPI on known services. For example, \fB-m\fR will detect both SSH and HTTPS on port 443.
.IP \fB-N\fR
Resolve IP addresses to hostnames.
.IP \fB-P\fR
Enable promiscuous mode.
.IP \fB-C\fR
Force color output even if output is not a tty.
.IP \fB-D\fR
Disable DPI. Only decode known services (DPI is enabled by default on unknown ports).
.IP "\fB-i \fIinterface\fR"
Specify the interface to listen on.
.IP "\fB-p \fIpcapfile\fR"
Process the given PCAP capture file instead of packets observed on the network.
.IP "\fB-s \fIsnaplen\fR"
Analyze at most the first \fIsnaplen\fR bytes of each TCP connection,
rather than the default of 1024.
.IP "\fB-f \fIservices\fR"
Load triggers from a \fIservices\fR file. The default is to use the built-in list.
.IP "\fB-M \fImagicfile\fR"
Load magic signatures from a file. The default is to use the built-in list. Cannot be used with \fB-t\fR. Deep Packet Inspection (DPI) is based on Unix magic files. This feature can be used to selectively enable DPI. For example, to capture SSH banners on any port, use:
.RS
dsniff -M <(echo "0 string SSH-2.0- ssh")
.RE
.IP "\fB-t \fItrigger\fR[,...]"
Load triggers from a comma-separated list, specified as
\fIport\fR/\fIproto\fR=\fIservice\fR (e.g., 80/tcp=http). Triggers set default decoders for well-known services. The default is to use the built-in list.
.IP "\fIexpression\fR"
Specify a tcpdump(8) filter expression to select traffic to sniff.
.LP
On a SIGHUP signal, \fBdsniff\fR will dump its current trigger table to
\fIdsniff.services\fR.
.SH ENVIRONMENT
The following environment variable can be set:
.LP
.B ENV_ARGS
A string of command-line options. Using this variable hides the options from the process list:
.RS
# ENV_ARGS="-i eth0 -v" ./dsniff
.RE
.SH "SEE ALSO"
arpspoof(8), libnids(3), services(5), magic(5)
.SH AUTHOR
.na
.nf
Dug Song <dugsong@monkey.org>
Skyper <skyper@thc.org>
.SH HISTORY
\fBdsniff\fR was written by Dug Song in 2000. It is currently maintained by
The Hacker's Choice (https://thc.org). It is available at https://github.com/hackerschoice/dsniff.
.SH BUGS
\fBdsniff\fR's automatic protocol detection feature is based on the
classic file(1) command by Ian Darwin and shares its historical
limitations and bugs.


debian/changelog
dsniff (2.4b1+debian-31) unstable; urgency=medium

  [ Marcos Fouces ]
  * Acknowledge Janitor commits.
  * Update d/copyright.

  [ Debian Janitor ]
  * Update standards version to 4.6.0, no changes needed.
  * Remove constraints unnecessary since buster (oldstable)
  * Update standards version to 4.6.1, no changes needed.

 -- Marcos Fouces <marcos@debian.org>  Fri, 04 Nov 2022 18:22:37 +0100

dsniff (2.4b1+debian-30) unstable; urgency=medium

  [ Samuel Henrique ]
  * Add salsa-ci.yml
  * Configure git-buildpackage for Debian

  [ Deb Janitor ]
  * Use secure URI in Homepage field.
  * Set debhelper-compat version in Build-Depends.
  * Fix day-of-week for changelog entries 2.4b1-10, 2.4b1-3.6, 2.4b1-3.5.

  [ Marcos Fouces ]
  * Bump to debhelper-compat 12.
  * Add Rules-Requires-Root: no in control file.
  * Bump to Standards-Version 4.5.0
  * Add modified 37_fix-lib-and-share-dirs.patch (Closes: #919631).
  * Remove -Wl,--as-needed linker flags as they are no more needed
    with Bullseye toolchain.
  * Update uploader email to @debian.org.
  * Add 38_fix-pcap_init.patch (Closes: #980588)
  * Bump to debhelper-compat level 13.
  * Bump to Standards-Version 4.5.1 (no changes required).
  * Build against libtirpc as a separate library.

 -- Marcos Fouces <marcos@debian.org>  Wed, 10 Feb 2021 18:26:44 +0100

dsniff (2.4b1+debian-29) unstable; urgency=medium

  [ Raphaël Hertzog ]
  * Update team maintainer address to
    Debian Security Tools <team+pkg-security@tracker.debian.org>
  * Update Vcs-Git and Vcs-Browser for the move to salsa.debian.org

  [ Samuel Henrique ]
  * d/copyright: Remove trailing whitespace and align author names

  [ Marcos Fouces ]
  * Update to dh 11
  * Remove trailing whitespaces.
  * Fix spelling errors spotted by Lintian.
  * Update copyright years.
  * Add proper Files-excluded field in d/copyright.
  * Bump to Standards-revision 4.2.1 (no changes required).
  * Use secure URL in d/watch file.
  * Bump to uscan version 4.

 -- Marcos Fouces <marcos.fouces@gmail.com>  Sun, 04 Nov 2018 22:22:12 -0200

dsniff (2.4b1+debian-28) unstable; urgency=medium

  [ Lukas Schwaighofer ]
  * fix implicit declarations compiler warning
    - switch to C99 uint64_t, remove now unnecessary xdr_u_int64_t function
    - include missing string.h for memset
  * bump Standards version to 4.0.0 (no change required)

 -- Lukas Schwaighofer <lukas@schwaighofer.name>  Fri, 21 Jul 2017 20:05:34 +0200

dsniff (2.4b1+debian-27) unstable; urgency=high

  * Fix a parallel FTBFS due to missing Makefile dependencies.  Thanks to
    Adrian Bunk (Closes: #869086).

 -- Lukas Schwaighofer <lukas@schwaighofer.name>  Thu, 20 Jul 2017 20:00:38 +0200

dsniff (2.4b1+debian-26) unstable; urgency=medium

  * Team upload.

  [ Marcos Fouces ]
  * Fix 27_arpspoof-allow-use-of-multiple-interface.patch
    Thanks to João Salavisa. Closes: #706766
  * Add 34_Add_CPPFLAGS.patch to build binaries with
    all default Debian compiler flags.

  [ Lukas Schwaighofer ]
  * no longer export CC, cross compiling works fine without on dh 10
  * avoid unnecessary linking by passing -Wl,--as-needed
  * pass prefix= to dh_auto_install instead of calling make directly
  * remove outdated information from README.Debian, document notable changes
    from upstream instead

  [ Raphaël Hertzog ]
  * Drop Recommends on libx11-6, it's already in Depends via
    shlibs:Depends.

 -- Raphaël Hertzog <hertzog@debian.org>  Thu, 15 Jun 2017 18:20:50 +0200

dsniff (2.4b1+debian-25) unstable; urgency=medium

  * debian/control: Added myself to Uploaders.
  * Make sure libmissing.a is built before PROGS (Closes: #860611).

 -- Lukas Schwaighofer <lukas@schwaighofer.name>  Wed, 19 Apr 2017 20:15:27 +0200

dsniff (2.4b1+debian-24) unstable; urgency=medium

  * Fix FTCBFS: Pass triplet-prefixed CC to configure.
    Thanks to Helmut Grohne (Closes: #852360).
  * Add four patches from Fedora:
    (Closes: #715646, #716355, #716457, #716458)
    + 29_libnet_name2addr4.patch
    + 30_pntohl_shift.patch
    + 31_sysconf_clocks.patch
    + 32_rpc_segfault.patch (Closes: #855869)
  * Polish, reorder and refresh patches.
  * Add 33_sshcrypto_DES.patch
  * Improve dep3 description in 33_sshcrypto_DES.patch
  (as request by Stretch release team)

 -- Marcos Fouces <mfouces@yahoo.es>  Mon, 13 Mar 2017 18:34:19 +0100

dsniff (2.4b1+debian-23) unstable; urgency=medium

  * Assign to pkg-security team (Closes: #847505)
  * Add 24_Fix-OpenSSL1.1.0-Build.patch.
    Thanks to Christoph Biedl. Closes (#828287)
  * Bump to dh compat level 10
  * Update copyright file
  * Bump to standards-revision 3.9.8
  * Fix copyright file
  * Add hardening options
  * Add 25_fix-spelling-errors.patch

 -- Marcos Fouces <mfouces@yahoo.es>  Tue, 06 Dec 2016 12:08:15 +0100

dsniff (2.4b1+debian-22.1) unstable; urgency=low

  * Non-maintainer upload.
  * Fix TDS decoder patch (Closes: #712648).

 -- Andrew Shadura <andrewsh@debian.org>  Sat, 30 Nov 2013 21:29:23 +0100

dsniff (2.4b1+debian-22) unstable; urgency=low

  * Acknowledge NMU.
     + Thanks Hilko Bengen
  * Bump standards to 3.9.3.
  * Added a patches:
     + Fix POP decoder in dsniff (Closes: #647583).
     + Add both communication partners in arpspoof (Closes: #650749).
     + Allow multiple targets to be imitated simultaniously (Closes: #650751).
     + arpspoof: allow the selection of source hw address (Closes: #650752).
     + Thanks Stefan Tomanek.

 -- William Vera <billy@billy.com.mx>  Tue, 06 Mar 2012 22:42:43 -0600

dsniff (2.4b1+debian-21.1) unstable; urgency=low

  * Non-maintainer upload.
  * Updated urlsnarf timestamp patch (Closes: #636202)

 -- Hilko Bengen <bengen@debian.org>  Tue, 06 Sep 2011 10:15:26 +0200

dsniff (2.4b1+debian-21) unstable; urgency=low

  * Changed patch system to quilt from dpatch.
  * Updated debian/README.source for quilt.
  * Removed lintian override.
  * Cleanup the debian/rules.
  * Build-Depend changed to libdb-dev (>= 4.7) (Closes: #548478).

dsniff.c
/*
 * dsniff.c
 *
 * Password sniffer, because DrHoney wanted one.
 *
 * This is intended for demonstration purposes and educational use only.
 *
 * Copyright (c) 2000 Dug Song <dugsong@monkey.org>
 *
 * $Id: dsniff.c,v 1.69 2001/03/15 08:33:03 dugsong Exp $
 */

#include "config.h"

#include <sys/types.h>

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <err.h>
#include <libnet.h>
#include <nids.h>
#include <pcap.h>

#include "options.h"
#include "pathnames.h"
#include "pcaputil.h"
#include "trigger.h"
#include "record.h"
#include "env2argv.h"

#define MAX_LINES	6
#define MIN_SNAPLEN	2048

int	Opt_client = 0;
int	Opt_debug = 0;
u_short	Opt_dns = 0;
int	Opt_magic = 1;
int	Opt_read = 0;
int	Opt_write = 0;
int	Opt_snaplen = MIN_SNAPLEN;
int	Opt_lines = MAX_LINES;
int Opt_verbose = 0;
int Opt_show_dups = 0;
int Opt_color = 0;

static void
usage(void)
{
	fprintf(stderr, "Version: " VERSION "\n"
"Usage: dsniff [-cdamDNPCv] [-i interface | -p pcapfile] [-s snaplen]\n"
"              [-f services] [-t trigger[,...]]\n"
"              [pcap filter]\n"
" -c         Half-duplex TCP stream assembly\n"
" -a         Show duplicates\n"
" -v         Verbose. Show banners. [-vv more verbose]\n"
" -d         Show debugging information\n"
" -D         Disable DPI. Only decode known ports.\n"
" -m         Force DPI also on known ports (e.g. ignore /etc/services).\n"
"            For example, -m will detect SSH on port 443 (https).\n"
" -C         Force color output even if not a TTY (disable color: dsniff|cat)\n"
" -N         Resolve IP addresses to hostname\n"
" -P         Enable promisc mode\n"
" -t <...>   Force a decoding method for a specific port/protocol.\n"
"            Example: Decode IMAP on port 8143: -t 8143/tcp=imap\n"
" -i <link>  Specify the interface to listen on\n"
" -p <file>  Read from pcap file\n"
" -s <len>   Analyze at most the first snaplen of each TCP connection [default: %d]\n"
#ifdef WITH_BERKELEY_DB
" -w <db>    Write sniffed sessions to db rather then printing them out\n"
" -r <db>    Read sniffed sessions from a db created with the -w option\n"
#endif
"\n" 
" Example:\n"
"   dsniff -i eth0 -C >log.txt\n", MIN_SNAPLEN);
	exit(1);
}

static void
sig_hup(int sig)
{
	trigger_dump();
}

static void
sig_die(int sig)
{
	record_close();
	exit(0);
}

static void
null_syslog(int type, int errnum, struct ip *iph, void *data)
{
}


static int get_all_ifaces(struct ifreq **, int *);
static unsigned int get_addr_from_ifreq(struct ifreq *);

int all_local_ipaddrs_chksum_disable()
{
	struct ifreq *ifaces;
	int ifaces_count;
	int i, ind = 0;
	struct nids_chksum_ctl *ctlp;
	unsigned int tmp;

	if (!get_all_ifaces(&ifaces, &ifaces_count))
		return -1;
	ctlp =
	    (struct nids_chksum_ctl *) malloc(ifaces_count *
					      sizeof(struct
						     nids_chksum_ctl));
	if (!ctlp)
		return -1;
	for (i = 0; i < ifaces_count; i++) {
		tmp = get_addr_from_ifreq(ifaces + i);
		if (tmp) {
			ctlp[ind].netaddr = tmp;
			ctlp[ind].mask = inet_addr("255.255.255.255");
			ctlp[ind].action = NIDS_DONT_CHKSUM;
			ind++;
		}
	}
	free(ifaces);
	nids_register_chksum_ctl(ctlp, ind);
}

/* helper functions for Example 2 */
unsigned int get_addr_from_ifreq(struct ifreq *iface)
{
	if (iface->ifr_addr.sa_family == AF_INET)
		return ((struct sockaddr_in *) &(iface->ifr_addr))->
		    sin_addr.s_addr;
	return 0;
}

static int get_all_ifaces(struct ifreq **ifaces, int *count)
{
	int ifaces_size = 8 * sizeof(struct ifreq);
	struct ifconf param;
	int sock;
	unsigned int i;

	*ifaces = malloc(ifaces_size);
	sock = socket(PF_INET, SOCK_DGRAM, IPPROTO_IP);
	if (sock <= 0)
		return 0;
	for (;;) {
		param.ifc_len = ifaces_size;
		param.ifc_req = *ifaces;
		if (ioctl(sock, SIOCGIFCONF, &param))
			goto err;
		if (param.ifc_len < ifaces_size)
			break;
		free(*ifaces);
		ifaces_size *= 2;
		ifaces = malloc(ifaces_size);
	}
	*count = param.ifc_len / sizeof(struct ifreq);
	close(sock);
	return 1;
      err:
	close(sock);
	return 0;
}

int
main(int argc, char *argv[])
{
	extern char *optarg;
	extern int optind;
	char *services, *savefile, *triggers, *magics;
	int c;

	env2argv(&argc, &argv);
	services = savefile = triggers = magics = NULL;
	
	if (isatty(STDOUT_FILENO))
		Opt_color = 1;

	nids_params.promisc = 0;

	while ((c = getopt(argc, argv, "PCvcdDaM:f:i:mNnp:r:s:t:w:h?V")) != -1) {
		switch (c) {
		case 'P':
			nids_params.promisc = 1;
			break;
		case 'C':
			Opt_color = 2; // FORCED
			break;
		case 'v':
			Opt_verbose = 1;
			break;
		case 'c':
			Opt_client = 1;
			break;
		case 'a':
			Opt_show_dups++;
			break;
		case 'd':
			Opt_debug++;
			break;
		case 'f':
			services = optarg;
			break;
		case 'M':
			magics = optarg;
			break;
		case 'i':
			nids_params.device = optarg;
			break;
		case 'D':
			Opt_magic = 0;
			break;
		case 'm':
			Opt_magic++;
			break;
		case 'N':
			Opt_dns = 1;
			break;
		case 'n':
			// compat mode
			Opt_dns = 0;
			break;
		case 'p':
			nids_params.filename = optarg;
			break;
		case 's':
			if ((Opt_snaplen = atoi(optarg)) == 0)
				usage();
			break;
		case 't':
			triggers = optarg;
			trigger_init_list(triggers);
			break;
#ifdef WITH_BERKELEY_DB
		case 'r':
			Opt_read = 1;
			savefile = optarg;
			break;
		case 'w':
			Opt_write = 1;
			savefile = optarg;
			break;
#endif
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;
	
	if (Opt_read && Opt_write)
		usage();
	if (Opt_write && (Opt_color < 2))
		Opt_color = 0; // Disable color for to DB.

	if (!record_init(savefile))
		err(1, "record_init");
	
	signal(SIGHUP, sig_hup);
	signal(SIGINT, sig_die);
	signal(SIGTERM, sig_die);

	if (Opt_read) {
		record_dump();
		record_close();
		exit(0);
	}

	if (argc != 0)
		nids_params.pcap_filter = copy_argv(argv);
	nids_params.scan_num_hosts = 0;
	nids_params.syslog = null_syslog;
	nids_params.n_tcp_streams = 8 * 1024;
	nids_params.tcp_workarounds = 1;
	
	if (!nids_init()) {
		record_close();
		errx(1, "nids_init: %s", nids_errbuf);
	}

	// HERE: Manual -t triggers have already been loaded.
	// Add DPI magics
	if (Opt_magic >= 1)
		trigger_init_magic(magics);
	if (Opt_magic >= 2)
		fprintf(stderr, "\
WARNING: DPI (-m) wont detect some protocols. Use -t to assist the DPI.\n\
Example: To decode IMAP on port 8143 use '-t 8143/tcp=imap'.");
	// Add known ports to triggers unless DPI is forced on all ports (-m)
	if (Opt_magic != 2)
		trigger_init_services(services);

	nids_register_ip(trigger_ip);
	nids_register_ip(trigger_udp);
		
	if (Opt_client) {
		nids_register_ip(trigger_tcp_raw);
		signal(SIGALRM, trigger_tcp_raw_timeout);
		alarm(TRIGGER_TCP_RAW_TIMEOUT);
	}
	else nids_register_tcp(trigger_tcp);
	
	if (nids_params.pcap_filter != NULL) {
		if (nids_params.filename == NULL) {
			warnx("listening on %s [%s]", nids_params.device,
		        nids_params.pcap_filter);
		}
		else {
			warnx("using %s [%s]", nids_params.filename,
		        nids_params.pcap_filter);
		}
	}
	else {
		if (nids_params.filename == NULL) {
			warnx("listening on %s", nids_params.device);
		}
		else {
			warnx("using %s", nids_params.filename);
		}
	}

	// all_local_ipaddrs_chksum_disable();
	// In Docker containers the chksum is often wrong even if not send from any
	// of the local NICs (local to the PHY but not visible by the container)....
	struct nids_chksum_ctl chksum_conf = { 0, 0, NIDS_DONT_CHKSUM };
	nids_register_chksum_ctl(&chksum_conf, 1);

	nids_run();
	
	/* NOTREACHED */
	
	exit(0);
}
